prody/utilities/catchall.py

```diff
@@ -7,13 +7,125 @@
     pass
 
+def calcModeClusters(similarities, n_clusters=0, linkage='all', method='tsne', cutoff=0.0, **kwargs):
+    """Perform clustering based on members of the *ensemble* projected into lower a reduced
+    dimension.
+    
+    :arg similarities: a matrix of similarities for each structure in the ensemble, such as
+                        RMSD-matrix, dynamics-based spectral overlap, sequence similarity
+    :type similarities: :class:`~numpy.ndarray`
+
+    :arg n_clusters: the number of clusters to generate. If **0**, will scan a range of 
+                        number of clusters and return the best one based on highest
+                        silhouette score. Default is **0**.
+    :type n_clusters: int
+
+    :arg linkage: if **all**, will test all linkage types (ward, average, complete,
+                    single). Otherwise will use only the one given as input. Default is
+                    **all**.
+    :type linkage: str
+
+    :arg method: if set to **spectral**, will generate a Kirchoff matrix based on the 
+                    cutoff value given and use that as input as clustering instead of
+                    the values themselves. Default is **tsne**.
+    :type method: str
+
+    :arg cutoff: only used if *method* is set to **spectral**. This value is used for 
+                    generating the Kirchoff matrix to use for generating clusters when
+                    doing spectral clustering. Default is **0.0**.
+    :type cutoff: float
+    """
+    
+    # Import necessary packages
+    try:
+        from sklearn.manifold import SpectralEmbedding
+        from sklearn.cluster import AgglomerativeClustering
+        from sklearn.metrics import silhouette_score
+        from sklearn.manifold import TSNE
+    except ImportError:
+        raise ImportError('need sklearn module')
+
+    # Check inputs to make sure are of valid types/values
+    if not isinstance(similarities, np.ndarray):
+        try:
+            import pandas as pd
+            if not isinstance(similarities, np.ndarray):
+                raise TypeError('similarities should be an instance of ndarray or DataFrame')
+        except:
+            raise TypeError('similarities should be an instance of ndarray or DataFrame')
+
+    dim = similarities.shape
+    if dim[0] != dim[1]:
+        raise ValueError('similarities must be a square matrix')
+
+    if n_clusters != 0:
+        if not isinstance(n_clusters, int):
+            raise TypeError('clusters must be an instance of int')
+        if n_clusters < 1:
+            raise ValueError('clusters must be a positive integer')
+        elif n_clusters > similarities.shape[0]:
+            raise ValueError('clusters can\'t be longer than similarities matrix')
+        nclusts = range(n_clusters,n_clusters+1)
+    else:
+        nclusts = range(2,10,1)
+
+    if linkage != 'all':
+        if not isinstance(method, str):
+            raise TypeError('linkage must be an instance of str')
```

Comment:
It might be good to let the user provide a list-like of methods

Reply:
That's a good idea, I've added that as an option.